### PR TITLE
fix: Avoid calling __class__ when describing values

### PR DESCRIPTION
**Pull request by dividedmind**
The described value could overload `__class__` to something which causes side effects.

This was the cause of the problem found by @symwell in https://github.com/getappmap/appmap-python/issues/165?notification_referrer_id=NT_kwDOAAyRVLE0NDI4ODc3ODA4OjgyMzYzNg#issuecomment-1295264871

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,70 +1,53 @@
...
-addons:
-  apt:
-    update: true
-    packages: [build-essential zlib1g-dev libncurses5-dev libgdbm-dev libnss3-dev libssl-dev libreadline-dev libffi-dev libsqlite3-dev wget libbz2-dev]
+language: python
+python:
+- "3.10"
+- "3.9"
+- "3.8"
+- "3.7"
```
Comment (apotterri) [inline .travis.yml:8]: The downside to this is that it puts us on the Travis maintainer's release schedule. The versions of python that are currently available on a worker are quite a bit behind.

FWIW, I'd verified that Travis was doing a good-enough job caching pyenv builds. We should discuss with the team whether we want to make this change.

**Review by apotterri — CHANGES_REQUESTED** [review]

File: pyproject.toml [hunk 2]
```diff
@@ -60,7 +60,6 @@ coverage = "^5.3"
 pytest-mock = "^3.5.1"
 SQLAlchemy = { version = "^1.4.11", python = "^3.7"}
 tox = "^3.22.0"
-tox-pyenv = "^1.1.0"
 tox-travis = ">=0.12"
```
[the remark this answers is not in the document]
Comment (dividedmind) [reply pyproject.toml:63]: Ah, yes, good point.

File: tox.ini [hunk 3]
```diff
@@ -21,6 +21,7 @@ commands =
     # Turn off recording while installing. It's not necessary, and the warning messages that come
     # out of the agent confuse poetry.
     env APPMAP=false poetry install -v
+    py310-web: poetry run pylint appmap
```
[the remark this answers is not in the document]
Comment (dividedmind) [reply tox.ini:24]: Travisrc suggested it was intended to run, but for some reason apparently it never did. I chose to add it and lower the code quality threshold for now. I'll create another issue to improve it, but I wouldn't like to see it creep down even lower.

File: appmap/test/data/appmap_testing/django_simplelazyobject.py [hunk 4]
```diff
@@ -0,0 +1,9 @@
+# pylint: disable=missing-module-docstring,missing-function-docstring
+
+from django.utils.functional import SimpleLazyObject
```
[the remark this answers is not in the document]
Comment (dividedmind) [reply appmap/test/data/appmap_testing/django_simplelazyobject.py:3]: Yes, it does. `SimpleLazyObject` is an example of a real-life code that exhibits this problem. (This is alluded to in the comment at https://github.com/getappmap/appmap-python/pull/188/files#diff-394d7efa89fced65f53ad7bc6fe8897a0c1fc46c9e737939483c6fc6c5366005R77 )

File: .travis.yml (hunk 1 above)
Comment (dividedmind) [reply .travis.yml:8]: I don't think it's that important unless we actually do find a problem with some minor release (and if that happens we can always just point-test that one in a separate and short build step). I think it makes more sense to use travis-provided pythons — it cuts down build time, drastically simplifies the CI configuration and enables build parallelism (sure, we make our own version matrix, but that complicates it even further). It's also more robust (as the build being suddenly broken — which led me to do this change — shows). I don't think we really need to live on the bleeding edge.

File: appmap/test/data/appmap_testing/django_simplelazyobject.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dividedmind) [reply appmap/test/data/appmap_testing/django_simplelazyobject.py:3]: But it's a good idea to also mention it in the commit message :)

[the remark this answers is not in the document]
Comment (apotterri) [reply appmap/test/data/appmap_testing/django_simplelazyobject.py:3]: Sorry, missed the reference in the other comment. Thanks.

File: .travis.yml (hunk 1 above)
Comment (apotterri) [reply .travis.yml:8]: ok

File: tox.ini (hunk 3 above)
[the remark this answers is not in the document]
Comment (apotterri) [reply tox.ini:24]: Yeah, the command line that was getting used to try to run it was wrong.

**Review by apotterri — APPROVED** [review]

**Comment by apotterri** [comment]
This is good to go, once the build is green, and the [deploy step](https://github.com/getappmap/appmap-python/pull/188/commits/e95fdacf34ee5f4b816c45e1aa8e04d79acd8d38#diff-6ac3f79fc25d95cd1e3d51da53a4b21b939437392578a35ae8cd6d5366ca5485R34) is fixed.

**Comment by apotterri** [comment]
> This was the cause of the problem found by @symwell in https://github.com/getappmap/appmap-python/issues/165

So, should this PR be linked to #165?

**Comment by symwell** [comment]
Were the saleor testcases re-run after this fix?  I'm running some of them now manually.

**Comment by dividedmind** [comment]
> > This was the cause of the problem found by @symwell in #165
> 
> So, should this PR be linked to #165?

Not as a fixer since it's not the only problem preventing saleor tests to pass.

**Comment by dividedmind** [comment]
> Were the saleor testcases re-run after this fix? I'm running some of them now manually.

I started to rerun it and it did fix some, but there are still some that fail. I intend to take a look at them next.

**Comment by dividedmind** [comment]
> This is good to go, once the build is green, and the [deploy step](https://github.com/getappmap/appmap-python/pull/188/commits/e95fdacf34ee5f4b816c45e1aa8e04d79acd8d38#diff-6ac3f79fc25d95cd1e3d51da53a4b21b939437392578a35ae8cd6d5366ca5485R34) is fixed.

Thanks! Deploy should be fixed now, could you just take a quick look at [the new deploy step](https://github.com/getappmap/appmap-python/pull/188/commits/f80b3f32ace9b0c4feefe8ccb3d4279a7a38a7f7#diff-6ac3f79fc25d95cd1e3d51da53a4b21b939437392578a35ae8cd6d5366ca5485R33) to make sure I didn't screw something up?

**Comment by apotterri** [comment]
Looks like everything's working properly -- I see that your changes got [deployed](https://app.travis-ci.com/github/getappmap/appmap-python/jobs/587595001#L426) as version 1.12.1 on [PyPI](https://pypi.org/project/appmap/1.12.1/)

**Pull request merged**